File: 234-palindrome-linked-list/palindrome-linked-list.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public bool IsPalindrome(ListNode head) {
        bool isPalindrome= true ;
        Stack<int> st = new Stack<int>();
        ListNode cur = head;
        while(cur != null){
            st.Push(cur.val);
            cur = cur.next;
            }
        cur = head;
        while (cur != null){
            if(cur.val != st.Pop()){
                isPalindrome = false;
                break;
            }
            cur=cur.next;
        }
        return isPalindrome;
    }
}

File: 234-palindrome-linked-list/palindrome-linked-list.cpp
class Solution {
public:
    ListNode  *getNode(int n,ListNode*head ){
        ListNode *cur = head;
        while (n--){
            cur = cur->next;
        }
        return cur;
    }
    ListNode* getMiddle(ListNode * head ){
        ListNode *slow = head;
        ListNode* fast = head;
        while (fast != NULL && fast->next!= NULL){
            slow = slow->next ;
            fast = fast->next->next;
        }
        return slow;
    }
    bool isPalindrome(ListNode* head) {
        ListNode * cur = getMiddle(head);
        ListNode* pre = NULL;
        ListNode * next = NULL;
        while(cur!=NULL){
            next = cur->next;
            cur->next = pre ;
            pre = cur ;
            cur = next;
        }
        ListNode * l = head ;
        ListNode *r = pre;
        while (r != NULL){
            if (r->val != l->val){
                return false ;
            } 
            l = l->next;
            r = r->next;
        } 
        return true;  
    }
};